### ISSUE TITLE
setImmediate not supported in most browsers

### ISSUE THREAD

**Issue by rossnanop** [mozilla-services/react-jsonschema-form#188]
### Prerequisites

- [X] I have read the [documentation](https://github.com/mozilla-services/react-jsonschema-form/blob/master/README.md#readme).

### Description

setImmediate function is causing validation to break because it is not implemented in most browsers (and looks like it will not become standard): 
https://developer.mozilla.org/en-US/docs/Web/API/Window/setImmediate

### Steps to Reproduce
I've tracked the error to this line: https://github.com/mozilla-services/react-jsonschema-form/blob/210a94421f4b034d46bac38cbc1f602358afc12a/src/utils.js#L381

1. Submit invalid form data or type in fields of a react-jsonschema-form 
2. See error in console in any browser from above docs

#### Expected behavior

No error.

#### Actual behavior
When submitting invalid data:
```bash
Uncaught ReferenceError: setImmediate is not defined 
setState @ index.js:15938 
Form._this.onSubmit @ index.js:13195
ReactErrorUtils.invokeGuardedCallback @ index.js:29632execute
Dispatch @ index.js:21812execute
DispatchesInOrder  @ index.js:21835execute
DispatchesAndRelease  @ index.js:21285execute
DispatchesAndReleaseTopLevel  @ index.js:21296for
EachAccumulated  @ index.js:35532
EventPluginHub.processEventQueue  @ index.js:21458
runEventQueueInBatch  @ index.js:29659
ReactEventEmitterMixin.handleTopLevel  @ index.js:29670
handleTopLevelImpl  @ index.js:29750
Mixin.perform  @ index.js:34978
ReactDefaultBatchingStrategy.batchedUpdates  @ index.js:28319
batchedUpdates  @ index.js:32671
ReactEventListener.dispatchEvent @ index.js:29827
```
When typing into fields:
```bash
Uncaught ReferenceError: setImmediate is not defined
setState @ index.js:15945
asyncSetState @ index.js:13947
(anonymous function) @ index.js:13909
onChange @ index.js:14741
LinkedValueUtils.executeOnChange @ index.js:22601
_handleChange @ index.js:27057
ReactErrorUtils.invokeGuardedCallback @ index.js:29639
executeDispatch @ index.js:21819
executeDispatchesInOrder @ index.js:21842
executeDispatchesAndRelease @ index.js:21292
executeDispatchesAndReleaseTopLevel @ index.js:21303
forEachAccumulated @ index.js:35539
EventPluginHub.processEventQueue @ index.js:21465
runEventQueueInBatch @ index.js:29666
ReactEventEmitterMixin.handleTopLevel @ index.js:29677
handleTopLevelImpl @ index.js:29757
Mixin.perform @ index.js:34985
ReactDefaultBatchingStrategy.batchedUpdates @ index.js:28326
batchedUpdates @ index.js:32678
ReactEventListener.dispatchEvent @ index.js:29834
```

### Version
react/react-dom 15.0.2
react-jsonschema-form 0.29.1

**Comment by n1k0** [comment]
That's surprising as it's polyfilled in the builds. I suspect you're building your own files from the package lib files, which doesn't expose a polyfill for it...

I know both browserify and webpack provide it by default, but maybe you're using something else? Or do use options like `--bare` or `no-builtins`? I'm really curious :)

Back to the issue, if we don't use `setImmediate`, what should we use instead in your opinion? `setTimeout(fn, 0)`? `requestAnimationFrame`?

**Comment by n1k0** [comment]
Could something be good enough?

```diff
--- a/src/utils.js
+++ b/src/utils.js
@@ -19,6 +19,10 @@ import ColorWidget from "./components/widgets/ColorWidget";
 
 const RE_ERROR_ARRAY_PATH = /\[\d+]/g;
 
+const setImmediate = global.setImmediate ||
+                     global.requestAnimationFrame ||
+                     global.setTimeout;
+
 const altWidgetMap = {
   boolean: {
     radio: RadioWidget,
```

**Comment by n1k0** [comment]
Also, there's [this](https://www.npmjs.com/package/setimmediate), and [this article](https://www.nczonline.net/blog/2013/07/09/the-case-for-setimmediate/) defending the setimmediate initiative. Hmm.

**Comment by rossnanop** [comment]
Yeah I was trying to avoid adding another dependency.  I like the solution you have above because it serves as a fallback if the global isn't available.  I personally don't have an opinion about using setImmediate, it just looks like there is a lot of friction about it and no guarantee it will be implemented.

Back to your earlier question.  I'm using browserify, but no options like --bare or no-builtins, so not sure why it is not being included.  I just used npm to grab react-jsonschema-form.

**Comment by rossnanop** [comment]
I am running ```npm rebuild --no-bin-links``` due to this issue: https://github.com/npm/npm/issues/9901

**Comment by n1k0** [comment]
@rossnanop could you please try building your project using the branch from #190 and tell me if that works for you? Thanks.

**Comment by rossnanop** [comment]
No longer get the error, but I am unable to type anything into the form.  It looks like the callback is just resetting the state to the default state?  This may be related to my issue (or a separate issue possibly).
```bash
index.js:38370 Warning: `value` prop on `input` should not be null. 
Consider using the empty string to clear the component or `undefined` for uncontrolled components.
```
React 15 now warns on null value props on input because they want to clarify intent.  In the BaseInput widget and a few other places, the check is:
```javascript 
typeof value === "undefined" ? "" : value
```
which doesn't cover value === null because typeof null === object.  I guess it just depends on if you want to allow people to pass null to react-jsonschema-form or not.  However, in some cases, you just want FormData to be empty if you don't have default values.

**Comment by n1k0** [comment]
@rossnanop I've updated #188 to use a decent polyfill to `setImmediate`, please check it out.

For your other issue, it's probably worth filing a new separate one.

**Comment by n1k0** [comment]
The fix has landed as part of [v0.30.0](https://github.com/mozilla-services/react-jsonschema-form/releases/tag/v0.30.0).
